sway/tree/container.c

```diff
@@ -1614,6 +1614,181 @@ void container_raise_floating(struct sway_container *con) {
+		} else if (ws1 != ws2) {
+			seat_set_focus_container(seat, focus == con1 ? con2 : con1);
+		} else {
+			seat_set_focus_container(seat, focus);
+		}
+	} else {
+		seat_set_focus_container(seat, focus);
+	}
+
+	if (root->fullscreen_global) {
+		seat_set_focus(seat,
+				seat_get_focus_inactive(seat, &root->fullscreen_global->node));
+	}
+}
+
+static void swap_places(struct sway_container *con1,
+		struct sway_container *con2) {
+	struct sway_container *temp = malloc(sizeof(struct sway_container));
+	temp->x = con1->x;
+	temp->y = con1->y;
+	temp->width = con1->width;
+	temp->height = con1->height;
+	temp->width_fraction = con1->width_fraction;
+	temp->height_fraction = con1->height_fraction;
+	temp->parent = con1->parent;
+	temp->workspace = con1->workspace;
+	bool temp_floating = container_is_floating(con1);
+
+	con1->x = con2->x;
+	con1->y = con2->y;
+	con1->width = con2->width;
+	con1->height = con2->height;
+	con1->width_fraction = con2->width_fraction;
+	con1->height_fraction = con2->height_fraction;
+
+	con2->x = temp->x;
+	con2->y = temp->y;
+	con2->width = temp->width;
+	con2->height = temp->height;
+	con2->width_fraction = temp->width_fraction;
+	con2->height_fraction = temp->height_fraction;
+
+	int temp_index = container_sibling_index(con1);
+	if (con2->parent) {
+		container_insert_child(con2->parent, con1,
+				container_sibling_index(con2));
+	} else if (container_is_floating(con2)) {
+		workspace_add_floating(con2->workspace, con1);
+	} else {
+		workspace_insert_tiling(con2->workspace, con1,
+				container_sibling_index(con2));
+	}
+	if (temp->parent) {
+		container_insert_child(temp->parent, con2, temp_index);
+	} else if (temp_floating) {
+		workspace_add_floating(temp->workspace, con2);
+	} else {
+		workspace_insert_tiling(temp->workspace, con2, temp_index);
+	}
+
+	free(temp);
+}
+
+void container_swap(struct sway_container *con1, struct sway_container *con2) {
+	if (!sway_assert(con1 && con2, "Cannot swap with nothing")) {
+		return;
+	}
+	if (!sway_assert(!container_has_ancestor(con1, con2)
+				&& !container_has_ancestor(con2, con1),
+				"Cannot swap ancestor and descendant")) {
+		return;
+	}
+
+	sway_log(SWAY_DEBUG, "Swapping containers %zu and %zu",
+			con1->node.id, con2->node.id);
+
+	bool scratch1 = con1->scratchpad;
+	bool hidden1 = container_is_scratchpad_hidden(con1);
+	bool scratch2 = con2->scratchpad;
+	bool hidden2 = container_is_scratchpad_hidden(con2);
+	if (scratch1) {
+		if (hidden1) {
+			root_scratchpad_show(con1);
+		}
+		root_scratchpad_remove_container(con1);
+	}
+	if (scratch2) {
+		if (hidden2) {
+			root_scratchpad_show(con2);
+		}
+		root_scratchpad_remove_container(con2);
+	}
+
+	enum sway_fullscreen_mode fs1 = con1->fullscreen_mode;
+	enum sway_fullscreen_mode fs2 = con2->fullscreen_mode;
+	if (fs1) {
+		container_fullscreen_disable(con1);
+	}
+	if (fs2) {
+		container_fullscreen_disable(con2);
+	}
+
+	struct sway_seat *seat = input_manager_current_seat();
```

Comment:
This is the only line changed in this commit. Otherwise these three functions were lifted directly from commands/swap.c